src/hooks/useModifyCharacters.ts

```diff
@@ -2,44 +2,196 @@ import { useReducer } from 'react'
 
 import { ModifiedCharacter } from '../api/starWars-types'
 
+export const CHARACTER_ACTION_TYPES = {
+  SORT: 'sort',
+  FILTER: 'filter',
+} as const
+
 export const SORT_FACTOR = {
   MASS: 'mass',
   HEIGHT: 'height',
 } as const
 
+export const FILTER_FACTOR = {
+  HOMEWORLD: 'homeworld',
+  GENDER: 'gender',
+} as const
+
 // "mass" | "height"
 export type SortFactor = typeof SORT_FACTOR[keyof typeof SORT_FACTOR]
 
+export type FilterFactor = typeof FILTER_FACTOR[keyof typeof FILTER_FACTOR]
+
+type FilterTargets = {
+  homelandTarget: string
+  genderTarget: string
+}
+
 type CharacterState = {
   originalCharacters: ModifiedCharacter[]
-  modifiedCharacters: ModifiedCharacter[]
-  selectedValue: SortFactor | ''
+  renderedCharacters: ModifiedCharacter[]
+  sortFactor: SortFactor | ''
+  homeworldFilters: string[]
+  genderFilters: string[]
+  filterTargets: FilterTargets
 }
 
-type CharacterAction = { type: 'sort'; sortFactor: CharacterState['selectedValue'] }
+type CharacterAction =
+  | {
+      type: 'sort'
+      sortFactor: CharacterState['sortFactor']
+    }
+  | { type: 'filter'; filterFactor: FilterFactor; filterTarget: string }
 
 const sort = (characters: ModifiedCharacter[], key: SortFactor) =>
   characters.sort((a, b) => (b[key] || 0) - (a[key] || 0))
```

Comment:
I prefer to generate a new array before execute `sort()`. I notice you understand that `sort()` is destructive and you pass a new array in some conditions. But, you sometimes assign a reference to a variable, for example, line 159 `let newRenderedCharacters = state.originalCharacters`. So far you can manage them but it may be confusing if another dev needs to modify.

Reply:
Good point. Errors from mutation with a reference can be hard to catch 👍